File: refactoring-movie/movie.dart
class Customer {
  Customer(this.name);

  String name;
  List<Rental> rentals = <Rental>[];

  void addRental(Rental rental) {
    rentals.add(rental);
  }

  String statement() {
    String result = "Rental record for $name\n";
    rentals.forEach((r) {
      result += r.describe();
    });

    result += "Amount owed is ${totalAmount()} \n";
    result += "You earned ${frequentRenterPoints} frequent renter points";
    return result;
  }

  double totalAmount() => rentals.fold(0, (total, r) => total += r.price());

  int frequentRenterPoints() =>
      rentals.fold(0, (total, r) => total += r.frequentRenterPoints());
}

class Rental {
  const Rental(this.movie, this.daysRented);

  final Movie movie;
  final int daysRented;

  double price() => movie.price(daysRented);

  int frequentRenterPoints() => movie.frequentRenterPoints(daysRented);

  String describe() => "\t ${movie.title} \t ${price()} \n";
}

abstract class Movie {
  const Movie(this.title);

  final String title;

  double price(int days);
  int frequentRenterPoints(int days);
}

class NewReleaseMovie implements Movie {
  const NewReleaseMovie(this.title);

  final String title;

  @override
  double price(int days) => days * 3.0;

  @override
  int frequentRenterPoints(int days) => days > 1 ? 2 : 1;
}

class RegularMovie implements Movie {
  const RegularMovie(this.title);

  final String title;

  @override
  double price(int days) => days > 2 ? (days - 2) * 1.5 : 2.0;

  @override
  int frequentRenterPoints(int days) => 1;
}

class ChildrenMovie implements Movie {
  const ChildrenMovie(this.title);

  final String title;

  @override
  double price(int days) {
    var total = 1.5;
    if (days > 3) {
      total += (days - 3) * 1.5;
    }
    return total;
  }

  @override
  int frequentRenterPoints(int days) => 1;
}

// class Movie {
//   static const int CHILDREN = 2;
//   static const int REGULAR = 0;
//   static const int NEW_RELEASE = 1;

//   String title;
//   int priceCode;

//   Movie(this.title, this.priceCode);
// }

// double totalAmount = 0;
// int frequentRenterPoints = 0;
// String result = "Rental record for $name\n";

// for (final rental in rentals) {
//   double amount = 0;
//   switch (rental.movie.priceCode) {
//     case Movie.REGULAR:
//       amount += 2;
//       if (rental.daysRented > 2) amount += (rental.daysRented - 2) * 1.5;
//       break;
//     case Movie.NEW_RELEASE:
//       amount += rental.daysRented * 3;
//       break;
//     case Movie.CHILDREN:
//       amount += 1.5;
//       if (rental.daysRented > 3) amount += (rental.daysRented - 3) * 1.5;
//       break;
//   }

//   // add frequent renter points
//   frequentRenterPoints++;

//   // add bonus for a two day new release rental
//   if (rental.movie.priceCode == Movie.NEW_RELEASE && rental.daysRented > 1)
//     frequentRenterPoints++;

//   // show figures for this rental
//   result += "\t ${rental.movie.title} \t $amount \n";

//   totalAmount += amount;
// }
// result += "Amount owed is $totalAmount \n";
// result += "You earned $frequentRenterPoints frequent renter points";

// return result;

File: refactoring-movie/movie.java
import java.util.ArrayList;
import java.util.List;

class Customer {

    private final String name;
    private final List<Rental> rentals = new ArrayList<Rental>();

    public Customer(final String name, List<Rental> rentals) {
        this.name = name;
        this.rentals = rentals;
    }

    public String getName() {
        return name;
    }

    public List<Rental> getRentals()
    {

    }

    //public void addRental(final Rental rental) {
      //  rentals.add(rental);
    //}
}

class CustomerStatementDomainService()
{
    Protected Customer customer;
    //private final List<Rental> rentals = new ArrayList<Rental>();

    public CustomerStatementDomainService(Customer customer) {
        this.customer = customer;
    }
}

interface calculateService
{
    public calculate();
}

class CustomerStatementCalculations implements calculateService
{
    private double totalAmount = 0;
    private int frequentRenterPoints = 0;
    private String result;
    Protected Customer customer;

    public CustomerStatementCalculations(Customer customer)
    {
        this.customer = customer;        
    }

    Protected execute()
    {
        this.initResult();
        this.calculate();
    }

    Protected initResult()
    {
        this.result = "Rental record for " + this.customer.getName() + "\n";
    }


    public calculate() {
        for (final Rental rental : this.customer.getRentals()) {
            double amount = 0;
            switch (rental.getMovie().getPriceCode()) {
                case Movie.REGULAR:
                    amount += 2;
                    if (rental.getDaysRented() > 2)
                        amount += (rental.getDaysRented() - 2) * 1.5;
                    break;
                case Movie.NEW_RELEASE:
                    amount += rental.getDaysRented() * 3;
                    break;
                case Movie.CHILDREN:
                    amount += 1.5;
                    if (rental.getDaysRented() > 3)
                        amount += (rental.getDaysRented() - 3) * 1.5;
                    break;
            }

            // add frequent renter points
            this.frequentRenterPoints++;
            // add bonus for a two day new release rental
            if (rental.getMovie().getPriceCode() == Movie.NEW_RELEASE && rental.getDaysRented() > 1)
                this.frequentRenterPoints++;

            // show figures for this rental
            

            totalAmount += amount;
        }

        result += "Amount owed is " + String.valueOf(totalAmount) + "\n";
        result += "You earned " + String.valueOf(frequentRenterPoints) + " frequent renter points";

        return result;
    }

    private addResultString()
    {
        this.result += "\t" + rental.getMovie().getTitle() + "\t" + String.valueOf(amount) + "\n";
    }

    public getResults()
    {
        return result;
    }
}

class Movie {
    public static final int CHILDREN = 2;
    public static final int REGULAR = 0;
    public static final int NEW_RELEASE = 1;

    private final String title;
    private int priceCode;

    public Movie(final String title, final int priceCode) {
        this.title = title;
        this.priceCode = priceCode;
    }

    public String getTitle() {
        return title;
    }

    public int getPriceCode() {
        return priceCode;
    }

    public void setPriceCode(final int priceCode) {
        this.priceCode = priceCode;
    }
}

class Rental {
    private final Movie movie;
    private final int daysRented;

    public Rental(final Movie movie, final int daysRented) {
        this.movie = movie;
        this.daysRented = daysRented;
    }

    public Movie getMovie() {
        return movie;
    }

    public int getDaysRented() {
        return daysRented;
    }
}